src/content/en/2021/webassembly.md

```diff
@@ -15,6 +16,461 @@ featured_stat_2: TODO
+        <td>`name`, `external_debug_info`</td>
+      </tr>
+      <tr>
+        <td><a href="https://superctf.com/ReactGodot.wasm">…/ReactGodot.wasm</a></td>
+        <td class="numeric">4,672,588</td>
+        <td>`name`</td>
+      </tr>
+      <tr>
+        <td><a href="https://ui.perfetto.dev/v18.0-591dd9336/trace_processor.wasm">…/v18.0-591dd9336/trace_processor.wasm</a></td>
+        <td class="numeric">2,079,991</td>
+        <td>`name`</td>
+      </tr>
+      <tr>
+        <td><a href="https://ui.perfetto.dev/v18.0-615704773/trace_processor.wasm">…/v18.0-615704773/trace_processor.wasm</a></td>
+        <td class="numeric">2,079,991</td>
+        <td>`name`</td>
+      </tr>
+      <tr>
+        <td><a href="https://unpkg.com/canvaskit-wasm@0.25.1/bin/profiling/canvaskit.wasm">…/canvaskit.wasm</a></td>
+        <td class="numeric">1,491,602</td>
+        <td>`name`</td>
+      </tr>
+    </tbody>
+  </table>
+  <figcaption>{{ figure_link(
+    caption="Largest custom sections.",
+    sheets_gid="179685689",
+    sql_file="large_custom_sections.sql"
+  ) }}</figcaption>
+</figure>
+
+All of those are almost exclusively the `name` section which contains function names for basic debugging. In fact, if we keep looking through the dataset, we can see that almost all of those custom sections contain just the debug information.
+
+### How much can we save by stripping debug info?
+
+While debug information is useful for local development, those sections can be hefty—they take up to 15 MiB before compression in the table above. If you want to be able to debug production issues users are experiencing, a better approach might be to strip the debug information out of the binary using `llvm-strip`, `wasm-strip` or `wasm-opt --strip-debug` before shipping, collect raw stacktraces and match them back to source locations locally, using the original binary.
+
+Let's check how much stripping this debug information would save us in combination with Brotli, vs. just Brotli from the previous step:
+
+{{ figure_markup(
+  caption="strip-debug + Brotli savings.",
+  description="Bar chart showing distribution of savings achieved by transformation described above. 0 percentile shows a tiny regression of 501 bytes, most others show no size change, 90% shows tiny improvements under 90 bytes on both desktop and mobile, but maximum achieved savings are at 2.4 MB on desktop and 1.3 MB on mobile.",
+  chart_url="https://docs.google.com/spreadsheets/d/e/2PACX-1vT6yhkn3lw148YQQHLoqA71NIsZLSSoBtgFmd_hRyhcmyPl2OpLyuOjUBk64I5DLE_grN8esL8oA3zt/pubchart?oid=1943000224&format=interactive",
+  sheets_gid="1401232418",
+  sql_file="sizes_and_savings.sql",
+  image="strip_br_savings.png"
+  )
+}}
+
+Most files in the archive already didn't have custom sections. In those that did, we could save up to 2.4 MiB / 1.3 MB for the largest Wasm binaries on desktop and mobile, which is a pretty noticeable improvement, especially on slow connections.
```

Comment:
I struggle how to apply the aforementioned chart changes to this graph since it shows so little EXCEPT for the extreme percentiles (0, and in particular 100).

So I rerun the stats using all 100 percentiles (using `UNNEST(GENERATE_ARRAY(0, 100)) AS percentile)) to see what it looked like, and put the results here: https://docs.google.com/spreadsheets/d/1IMa2SbdQgshb4pGWF1KOh9s4zMtLbRymWZGYjdaatXY/edit#gid=50034228

The produces this graph:

![image](https://user-images.githubusercontent.com/10931297/142759751-7043547a-c47a-440e-bbd2-1455c5dbc3e0.png)

You can see there really is nothing until right at the end, when it's a handful of (badly optimised?) examples.

Here's the raw data:

percentile | desktop | mobile
-- | -- | --
0 | -501 | -501
1 | 0 | 0
2 | 0 | 0
... | ... | ...
87 | 0 | 0
88 | 0 | 0
89 | 0 | 16
90 | 22 | 89
91 | 103 | 116
92 | 122 | 1,030
93 | 1,528 | 3,397
94 | 20,103 | 42,470
95 | 55,084 | 55,084
96 | 96,876 | 93,179
97 | 120,361 | 120,358
98 | 120,361 | 120,361
99 | 252,733 | 133,527
100 | 2,486,838 | 1,291,316


Should we just not have a graph and instead have a Big Number and/or discussion?
Or a table
Does this extra information change anything you want to say here?

Reply:
> Does this extra information change anything you want to say here?

Not really, that's part of what I wanted to show - that most toolchains already strip those sections, however in the few that don't, manually stripping custom sections can save even more.

Reply:
> I was considering making a graph over only unstripped modules

Maybe it would help to add such graph as an extra here though, and explain why I'm doing it? That could work better, because I agree that the extreme is a bit different / misleading in this case.

Reply:
Ok, please convert this graph into the same style. I'll then use it to show that most files already don't have custom sections, and I will add another graph right underneath to show savings over only files that did have those sections.

Reply:
Done, committed and staging update: https://20211121t105115-dot-webalmanac.uk.r.appspot.com/en/2021/webassembly#fig-18

Though not showing much when converted to kilobytes:

![image](https://user-images.githubusercontent.com/10931297/143062207-4c09abf5-bfe0-49fb-80e5-313115c10a7d.png)

Should we revert to bytes for this one? Though that feels a little bit of a con just to inflate the numbers... 😀

![image](https://user-images.githubusercontent.com/10931297/143062423-3377292a-1aa6-49af-96a4-a2bea5b0e621.png)

Reply:
Done: https://20211121t105115-dot-webalmanac.uk.r.appspot.com/en/2021/webassembly#fig-15

Pretty boring graph though:

![image](https://user-images.githubusercontent.com/10931297/143063053-ffdaab71-3d4a-46d7-92b9-8135c079cc64.png)

Reply:
That's okay, as I mentioned, it's still useful to show the zeroes in this case and I'll add another graph below :)

Looking at this and the other graph though, I'm less sure that converting to KB was a good idea now... Aside from initial sizes, all differences are not that big and might be actually easier to read in bytes.

Reply:
Here it is in bytes:

![image](https://user-images.githubusercontent.com/10931297/143067089-8f4290df-94d7-4136-a122-45121cc4284d.png)

Though to be honest I'm not sure 22 or even 89 bytes saving is much to shout about! 😀

Reply:
You know what, let's just add your 100 percentile scatter plot here instead, and I'll use it as a basis to explain the proper percentile graph over files with custom sections.

Reply:
On the plus side, it's another graph type - variety is the spice of live! 🌶

Here's the details:
- chart_url: https://docs.google.com/spreadsheets/d/e/2PACX-1vT6yhkn3lw148YQQHLoqA71NIsZLSSoBtgFmd_hRyhcmyPl2OpLyuOjUBk64I5DLE_grN8esL8oA3zt/pubchart?oid=852876683&format=interactive
- sheets_gid: 50034228
- sql_file: sizes_and_savings_100.sql

Will update the image and add the SQL before merging.

Reply:
I've gathered percentiles over files with custom sections and added a new graph meanwhile:
![image](https://user-images.githubusercontent.com/557590/143073173-4eb6d8b6-e3c5-407c-a943-6e1027b7b856.png)